internal/sourceanalysis/sourceanalysis.go

```diff
@@ -0,0 +1,57 @@
+package sourceanalysis
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/google/osv-scanner/internal/govulncheckshim"
+	"github.com/google/osv-scanner/internal/output"
+	"github.com/google/osv-scanner/pkg/models"
+	"golang.org/x/exp/slices"
+)
+
+// vulnsFromAllPkgs returns a map of IDs to GroupInfo pointers, and the extracted list of vulnerabilities
+func vulnsFromAllPkgs(pkgs []models.PackageVulns) (map[string]*models.GroupInfo, []models.Vulnerability) {
+	idToGroupMap := map[string]*models.GroupInfo{}
+	vulnList := []models.Vulnerability{}
+	for _, pv := range pkgs {
+		for _, vuln := range pv.Vulnerabilities {
+			groupIdx := slices.IndexFunc(pv.Groups, func(g models.GroupInfo) bool {
+				return slices.Contains(g.IDs, vuln.ID)
+			})
+
+			idToGroupMap[vuln.ID] = &pv.Groups[groupIdx]
+			vulnList = append(vulnList, vuln)
+		}
+	}
+
+	return idToGroupMap, vulnList
+}
+
+// DoSourceAnalysis runs the language specific analyzers on the code given packages and source info
+func DoSourceAnalysis(r *output.Reporter, source models.SourceInfo, pkgs []models.PackageVulns) {
```

Comment:
nit: rename this to just "Run" to avoid stutter since the package name is already "sourceanalysis".